ddl/ddl_tiflash_api.go

```diff
@@ -57,22 +57,156 @@ type TiFlashReplicaStatus struct {
 	IsPartition    bool
 }
 
+// TiFlashTick is type for backoff threshold.
+type TiFlashTick float64
+
+// PollTiFlashBackoffElement records backoff for each TiFlash Table.
+// When `Counter` reached `Threshold`, `Threshold` shall grow.
+type PollTiFlashBackoffElement struct {
+	Counter      int
+	Threshold    TiFlashTick
+	TotalCounter int
+}
+
+// NewPollTiFlashBackoffElement initialize backoff element for a TiFlash table.
+func NewPollTiFlashBackoffElement() *PollTiFlashBackoffElement {
+	return &PollTiFlashBackoffElement{
+		Counter:      0,
+		Threshold:    PollTiFlashBackoffMinTick,
+		TotalCounter: 0,
+	}
+}
+
+// PollTiFlashBackoffContext is a collection of all backoff states.
+type PollTiFlashBackoffContext struct {
+	MinTick TiFlashTick
+	MaxTick TiFlashTick
+	// Capacity limits tables a backoff pool can handle, in order to limit handling of big tables.
+	Capacity int
+	Rate     TiFlashTick
+	elements map[int64]*PollTiFlashBackoffElement
+}
+
+// NewPollTiFlashBackoffContext creates an instance of PollTiFlashBackoffContext.
+func NewPollTiFlashBackoffContext(MinTick, MaxTick TiFlashTick, Capacity int, Rate TiFlashTick) (*PollTiFlashBackoffContext, error) {
+	if MaxTick < MinTick {
+		return nil, fmt.Errorf("`MaxTick` should always be larger than `MinTick`")
+	}
+	if MinTick < 1 {
+		return nil, fmt.Errorf("`MinTick` should not be less than 1")
+	}
+	if Capacity < 0 {
+		return nil, fmt.Errorf("negative `Capacity`")
+	}
+	if Rate <= 1 {
+		return nil, fmt.Errorf("`Rate` should always be larger than 1")
+	}
+	return &PollTiFlashBackoffContext{
+		MinTick:  MinTick,
+		MaxTick:  MaxTick,
+		Capacity: Capacity,
+		elements: make(map[int64]*PollTiFlashBackoffElement),
+		Rate:     Rate,
+	}, nil
+}
+
 // TiFlashManagementContext is the context for TiFlash Replica Management
 type TiFlashManagementContext struct {
 	TiFlashStores             map[int64]helper.StoreStat
 	HandlePdCounter           uint64
 	UpdateTiFlashStoreCounter uint64
 	UpdateMap                 map[int64]bool
+	Backoff                   *PollTiFlashBackoffContext
+}
+
+// Tick will first check increase Counter.
+func (b *PollTiFlashBackoffContext) Tick(ID int64) (bool, bool, int) {
+	e, ok := b.Get(ID)
+	if !ok {
+		return false, false, 0
+	}
+	growed := e.MaybeGrow(b)
+	e.Counter += 1
+	e.TotalCounter += 1
+	return growed, true, e.TotalCounter
+}
+
+func (e *PollTiFlashBackoffElement) limit() int {
+	return int(e.Threshold)
+}
+
+// NeedGrow returns if we need to grow
+func (e *PollTiFlashBackoffElement) NeedGrow() bool {
+	return e.Counter >= e.limit()
+}
+
+func (e *PollTiFlashBackoffElement) doGrow(b *PollTiFlashBackoffContext) {
+	defer func() {
+		e.Counter = 0
```

Comment:
In this function only one place need to return. So I think we can put it in [line154](https://github.com/pingcap/tidb/pull/32317/files#diff-dca9a62898accc180490d462575cc0eae471c50e76c7348373075d5944addd93R154) directly.